test/util_test.go

```diff
@@ -92,8 +147,83 @@ func startDqlite(ctx context.Context, tb testing.TB) endpoint.Config {
 		app.Close()
 	})
 
-	return endpoint.Config{
-		Listener: fmt.Sprintf("unix://%s/listen.sock", dir),
+	db, err := app.Open(ctx, "k8s")
+	if err != nil {
+		tb.Fatal(err)
+	}
+
+	return &endpoint.Config{
+		Listener: fmt.Sprintf("unix://%s/kine.sock", dir),
 		Endpoint: fmt.Sprintf("dqlite://k8s?driver-name=%s", app.Driver()),
+	}, db
+}
+
+func setupScenario(ctx context.Context, db *sql.DB, prefix string, numInsert, numUpdates, numDeletes int) error {
+	t, err := db.BeginTx(ctx, nil)
+	if err != nil {
+		return err
+	}
+	defer t.Rollback()
+
+	insertManyQuery := `
+WITH RECURSIVE gen_id AS(
+	SELECT 1 AS id
+
+	UNION ALL
+
+	SELECT id + 1
+	FROM gen_id
+	WHERE id < ?
+), revision AS(
+	SELECT COALESCE(MAX(id), 0) AS base
+	FROM kine
+)
+INSERT INTO kine(
+	id, name, created, deleted, create_revision, prev_revision, lease, value, old_value
+)
+SELECT id + revision.base, ?||'/'||id, 1, 0, id + revision.base, 0, 0, 'value-'||id, NULL
+FROM gen_id, revision`
+	if _, err := t.ExecContext(ctx, insertManyQuery, numInsert, prefix); err != nil {
+		return err
 	}
+
+	updateManyQuery := fmt.Sprintf(`
+INSERT INTO kine(
+	name, created, deleted, create_revision, prev_revision, lease, value, old_value
+)
+SELECT kv.name, 0, 0, kv.create_revision, kv.id, 0, 'new-'||kv.value, kv.value
+FROM kine AS kv
+JOIN (
+	SELECT MAX(mkv.id) as id
+	FROM kine mkv
+	WHERE  ?||'/' <= mkv.name AND mkv.name < ?||'0'
+	GROUP BY mkv.name
+) maxkv ON maxkv.id = kv.id
+WHERE kv.deleted = 0
+ORDER BY kv.name
+LIMIT %d`, numUpdates)
```

Comment:
I think `numUpdates` slightly gets lost at the end of this line as the queries spans multiple lines. Perhaps this could be on a newline?